tests/Integration/Schema/Directives/InheritsDirectiveTest.php

```diff
@@ -0,0 +1,154 @@
+            }
+        ' . self::PLACEHOLDER_QUERY);
+
+        $childType = $schema->getType('ChildType');
+
+        $this->assertNotNull($childType->getField('attribute_1'));
+        $this->assertNotNull($childType->getField('attribute_2'));
+        $this->assertNotNull($childType->getField('attribute_3'));
+    }
+
+    public function testSchemaOrderIsInsignificant(): void
+    {
+        $schema = $this->buildSchema(/**@lang GraphQL */ '
+            type ChildType @inherits(from: "ParentType") {
+                attribute_3: String
+            }
+
+            type ParentType @inherits(from: "GrandParentType") {
+                attribute_2: String
+            }
+
+            type GrandParentType {
+                attribute_1: String
+            }
+        ' . self::PLACEHOLDER_QUERY);
+
+        $childType = $schema->getType('ChildType');
+
+        $this->assertNotNull($childType->getField('attribute_1'));
+        $this->assertNotNull($childType->getField('attribute_2'));
+        $this->assertNotNull($childType->getField('attribute_3'));
+    }
+
+    public function testChildOverridesFields(): void
+    {
+        $schema = $this->buildSchema(/* @lang GraphQL */ '
+            type ParentType {
+                attribute_1: String
+                attribute_2: Int
+            }
+
+            type ChildType @inherits(from: "ParentType") {
+                attribute_1: Int
+                attribute_2: String
+            }
+        ' . self::PLACEHOLDER_QUERY);
+
+        $childType = $schema->getType('ChildType');
+
+        $stringType = $schema->getType('String');
+        $intType = $schema->getType('Int');
+
+        $this->assertSame($intType, $childType->getField('attribute_1')->getType());
+        $this->assertSame($stringType, $childType->getField('attribute_2')->getType());
+    }
+
+    public function testChildAttributesShouldNotBeAddedParent(): void
+    {
+        $this->expectException(\GraphQL\Error\InvariantViolation::class);
+
+        $schema = $this->buildSchema(/* @lang GraphQL */ '
+            type ParentType {
+                attribute_1: String
+                another_attribute: String
+            }
+            
+            type ChildType @inherits(from: "ParentType") {
+                new_attribute: String
+            }
+        ' . self::PLACEHOLDER_QUERY);
+
+        $parentType = $schema->getType('ParentType');
+        $parentType->getField('new_attribute');
+    }
+
+    public function testInheritedTypeHasToMatch(): void
+    {
+        $this->expectException(\Nuwave\Lighthouse\Exceptions\DefinitionException::class);
+
+        $this->buildSchema(/*  @lang GraphQL */ '
+            input ParentType {
+                attribute_1: String
+            }
+            
+            type ChildType @inherits(from: "ParentType") {
+                attribute_2: String
+            }
+        ' . self::PLACEHOLDER_QUERY);
+    }
+
+    public function testForbidsInheritanceCycle(): void
```

Comment:
This test highlights another issue that we need to solve. Depending on the implementation, cyclic inheritance would either lead to nonsensical/incomplete results or endless loops. We need to detect cycles before resolving the inheritance tree.

Reply:
https://github.com/nuwave/lighthouse/pull/1895/commits/4b752906da64675b5bae7da88677a2c6855c2a5f

Just took care of this.

Reply:
Thanks, but that implementation is not sufficient - see [`fd894da` (#1895)](https://github.com/nuwave/lighthouse/pull/1895/commits/fd894dabd336bc8c90cd7d40da2f5c29d2827efa)

Reply:
https://github.com/webonyx/graphql-php/blob/master/src/Validator/Rules/NoFragmentCycles.php is doing a similar cycle detection and might give you an idea how to properly implement this.